client/components/combobox.jsx

```diff
@@ -1,128 +1,166 @@
-		return {
-			className   : '',
-			trigger     : 'hover',
-			default     : '',
-			placeholder : '',
-			autoSuggest : {
-				clearAutoSuggestOnClick : true,
-				suggestMethod           : 'includes',
-				filterOn                : []  // should allow as array to filter on multiple attributes, or even custom filter
-			},
-		};
-	},
-	getInitialState : function() {
-		return {
-			showDropdown : false,
-			value        : '',
-			options      : [...this.props.options],
-			inputFocused : false
+const Combobox = ({ onSelect, onEntry, autoSuggest = { filterOn: ['data-value'], suggestMethod: 'includes', clearAutoSuggestOnClick: false }, ...props })=>{
+	const [inputValue, setInputValue] = useState(props.value || '');
+	const [showDropdown, setShowDropdown] = useState(false);
+	const [inputFocused, setInputFocused] = useState(false);
+	const [currentOption, setCurrentOption] = useState(-1);
+	const [filteredOptions, setFilteredOptions] = useState(React.Children.toArray(props.children));
+	const inputRef = useRef(null);
+	const optionRefs = useRef([]);
+	const componentRef = useRef(null);
+
+	useEffect(()=>{
+		const handleClickOutside = (evt)=>{
+			if(showDropdown && componentRef.current && !componentRef.current.contains(evt.target)) {
+				setShowDropdown(false);
+			}
 		};
-	},
-	componentDidMount : function() {
-		if(this.props.trigger == 'click')
-			document.addEventListener('click', this.handleClickOutside);
-		this.setState({
-			value : this.props.default
-		});
-	},
-	componentWillUnmount : function() {
-		if(this.props.trigger == 'click')
-			document.removeEventListener('click', this.handleClickOutside);
-	},
-	handleClickOutside : function(e){
-		// Close dropdown when clicked outside
-		if(this.refs.dropdown && !this.refs.dropdown.contains(e.target)) {
-			this.handleDropdown(false);
+
+		document.addEventListener('pointerdown', handleClickOutside);
+		return ()=>{document.removeEventListener('pointerdown', handleClickOutside);};
+	}, [showDropdown]);
+
+	useEffect(()=>{
+		onSelect(inputValue);
+		// handleInputChange({ target: { value: inputValue } });
+	}, [inputValue]);
+
+	useEffect(()=>{
+		if(currentOption >= 0 && optionRefs.current[currentOption]) {
+			optionRefs.current[currentOption].focus();
 		}
-	},
-	handleDropdown : function(show){
-		this.setState({
-			showDropdown : show,
-			inputFocused : this.props.autoSuggest.clearAutoSuggestOnClick ? show : false
-		});
-	},
-	handleInput : function(e){
-		e.persist();
-		this.setState({
-			value        : e.target.value,
-			inputFocused : false
-		}, ()=>{
-			this.props.onEntry(e);
+	}, [currentOption]);
+
+	const handleInputChange = (evt)=>{
+		const newValue = evt.target.value;
+		setInputValue(newValue);
+		setCurrentOption(-1);
+
+		const filtered = React.Children.toArray(props.children).filter((option)=>{
+			return autoSuggest.filterOn.some((filterAttr)=>{
+				return option.props[filterAttr]?.toLowerCase().startsWith(newValue.toLowerCase());
+			});
```

Comment:
I believe we can make this using `some`, making it more concise without losing readability, make the change if you like

```suggestion
const filtered = props.children.filter(({ props }) =>
  autoSuggest.filterOn.some(filterAttr =>
    props[filterAttr]?.toLowerCase().startsWith(newValue.toLowerCase())
  )
);

```

Reply:
This is nearly the exact same as what I have unless I’m missing something or don’t know how to read a diff. 

The React.Children.toArray is important before there may be only a single child, not an array, so this would convert it to an array. Or, it could have multiple children but just not passed as an array. 

But regarding `some`, pretty sure I’m already using it the exact same way?

Reply:
basically, alright, then this is resolved. Did not think of the single element possibility